### PR TITLE
Adding value_and_grad to mlir. Continued work from #622

### PR DESCRIPTION
**Pull request by paul0403**
**Context:**  Adding value_and_grad to mlir. Continued work from #622

**Description of the Change:** frontend and mlir tests for `value_and_grad` needs to be added

**Benefits:** `catalyst.value_and_grad` can now be used on a qjit-ted function or qnode

**Possible Drawbacks:**

**Related GitHub Issues:**

[sc-55116]

### CODE REVIEW

**Comment by paul0403** [comment]
Changes from @rauletorresc 's previous work #622:

- Added a new optional string attribute `requester` to `BackpropOp`. This is to be set to `"GradOp"` when a `BackpropOp`is created from `HybridGradientLowering::matchAndRewrite()`, and to `"ValueAndGradOp"` when from `HybridValueAndGradientLowering::matchAndRewrite()`. Based on who created the backprop, in `ConversionPatterns.cpp`, `bool dupNoNeed` can be appropriately set. 
- This optional attribute does not break any existing tests: there are many instances in the existing test suites where `gradient.backprop` operations don't have this new attribute (e.g. https://github.com/PennyLaneAI/catalyst/blob/8330b7529a19ea1128830c2cb4b8f1c67efd5342/mlir/test/Gradient/BufferizationTest.mlir#L41)
- In the test where the `ConversionPatterns.cpp` actually takes place, i.e. `ConversionBackprop.mlir`,  this new attribute is added, once `"GradOp"`, once `"ValueAndGradOp"`, and once `"helloworld"` (to test against any possible future uses in other cases)

- (TODO) add frontend tests for `catalyst.value_and_grad`

**Comment by paul0403** [comment]
reminder to do changelog!

**Comment by paul0403** [comment]
Reminder to add value and grad to the grad op interface

**Comment by rauletorresc** [comment]
LGTM, however I will prefer @dime10 to approve, as I consider I might lack enough background

File: mlir/include/Gradient/IR/GradientOps.td
Comment (dime10) [inline mlir/include/Gradient/IR/GradientOps.td:144]: I think rather than specifying who invoked it, which requires one additional logic hop to determine what we actually want, let's go straight to the desired property `results_needed`, `discard_results`, etc.

Comment (paul0403) [reply mlir/include/Gradient/IR/GradientOps.td:144]: That makes sense, I'll update it

Comment (paul0403) [reply mlir/include/Gradient/IR/GradientOps.td]: done

File: mlir/lib/Gradient/IR/GradientOps.cpp
Comment (dime10) [inline mlir/lib/Gradient/IR/GradientOps.cpp:245]: `ps` and `adj` are no longer an option in the grad right?

**Review by dime10 — COMMENTED** [review]
Nice work 💯 Happy to approve after these two changes:

File: mlir/lib/Gradient/Transforms/GradMethods/HybridGradient.cpp
Comment (paul0403) [reply mlir/lib/Gradient/Transforms/GradMethods/HybridGradient.cpp]: done

File: mlir/lib/Gradient/IR/GradientOps.cpp
Comment (paul0403) [reply mlir/lib/Gradient/IR/GradientOps.cpp]: done

File: mlir/lib/Gradient/Transforms/GradMethods/HybridGradient.cpp
Comment (paul0403) [reply mlir/lib/Gradient/Transforms/GradMethods/HybridGradient.cpp:329]: I will try to factor out some common attributes with the new gradopinterface.

Comment (paul0403) [reply mlir/lib/Gradient/Transforms/GradMethods/HybridGradient.cpp:329]: This is corect

Comment (paul0403) [reply mlir/lib/Gradient/Transforms/GradMethods/HybridGradient.cpp]: done, the function signature is now the same as the original one, only taking in the new attribute for whether to keep the results

**Review by dime10 — APPROVED** [review]
🎉

File: frontend/test/pytest/test_gradient.py
Comment (josh146) [inline frontend/test/pytest/test_gradient.py:258]: Hey @paul0403, sorry I didn't see this in advance before merging! I just wanted to check that this works as well when the input is an `n-dim` tensor? In this case, the output of `value_and_grad` should be

```python
(scalar, params_shape)
```

Comment (paul0403) [reply frontend/test/pytest/test_gradient.py:258]: I just tried and actually it doesn't; however, it's not an issue with `value_and_grad`, as `grad` has the same issue. I will open an issue tonight or tomorrow.

Comment (paul0403) [reply frontend/test/pytest/test_gradient.py:258]: https://github.com/PennyLaneAI/catalyst/issues/841

**Pull request merged**
